### ISSUE TITLE
To avoid costly errors, validate features when they're added

### ISSUE THREAD

**Issue by softwaredoug** [o19s/elasticsearch-learning-to-rank#84]
If you have valid JSON, but a semantically invalid query (for example it's easy to add "query") like below when the plugin expects things at the "match" level:

```
 {
      "name": "user_rating",
      "type": "feature",
      "feature": {
         "name": "user_rating",
         "params": [],
         "template_language": "mustache",
         "template": {
            "query": {
               "match": {
                  "title": "{{keywords}}"
               }
            }
         }
      }
   }
```

Say this is feature "1" and you've built up a large feature set of 100 queries. Then you go to execute the feature set. You'll get a bug that the sltr query could not be executed.

As feature sets are append-only your whole feature set is screwed up and you have to start over. Is there any way to validate this is a valid query (ie attempt to parse it) before letting the feature be cerated

**Comment by nomoa** [comment]
I think two different levels of validations we could do:

- fully validation will require an index that has all the mappings required to run the queries. Would that work if the user has to provide the "test index" the query will be run into for testing?
- Simple json validation, we could validate that the query will pass various query builders but not validate everything that is related to mapping and field types/options. Problem is that I need to have he parameters replaced.

**Comment by softwaredoug** [comment]
Can we include an optional (perhaps mandatory) block for testing the feature that includes

- Index for testing
- Sample arguments for the template?

For example, when you create a feature, you might append a `test` block like so:

```
 {
      "name": "user_rating",
      "type": "feature",
      "feature": {
         "name": "user_rating",
         "params": ["keywords"],
         "test": {
             "keywords": "rambo",
             "index": "tmdb" 
         }
         "template_language": "mustache",
         "template": {
            "query": {
               "match": {
                  "title": "{{keywords}}"
               }
            }
         }
      }
   }
```

**Comment by softwaredoug** [comment]
Closed by #89
